lib/logstash/inputs/s3.rb

```diff
@@ -367,19 +371,18 @@ def ignore_filename?(filename)
     end
   end
 
-  def process_log(queue, key)
-    @logger.debug("Processing", :bucket => @bucket, :key => key)
-    object = @s3bucket.object(key)
+  def process_log(queue, log)
+    @logger.debug("Processing", :bucket => @bucket, :key => log.key)
+    object = @s3bucket.object(log.key)
 
-    filename = File.join(temporary_directory, File.basename(key))
+    filename = File.join(temporary_directory, File.basename(log.key))
     if download_remote_file(object, filename)
       if process_local_log(queue, filename, object)
-        lastmod = object.last_modified
```

Comment:
we're no longer storing the `last_modified` returned on the last call to retrieve the bucket, which is fine if ...
we would to assert that "listed" last_modified == as object last_modified and if that fails we should rather not process the given file atm and leave it for the next iteration (with some logging about what happened)?

would that make sense, to account for files being modified between listing and downloading?

Reply:
Adding log to ensure the same last_modified and the number of files being processed vs listed are good. These should help debug support case.

Leaving different last_modified files to the next iteration is a bit concerning. If a file is updated regularly and unluckily always be discovered beyond cutoff time, this file will be leftover for many iterations. It is an unlucky edge case. If duplication is not a concern, I would rather process it earlier. Do you think a file being processed multiple times is a concern?

Reply:
assert last_modified and log it should be enough for support. Counting the listed vs processed is kind of the same unless it is exposed to metrics

Reply:
> If a file is updated regularly and unluckily always be discovered beyond cutoff time, this file will be leftover for many iterations. 

just a summary - suppose a user is constantly updating an S3 bucket causing processing to be delayed.
current behavior: updates get processed but unpredictably - we do not guarantee what version of the file we download after an update but there's a guarantee never to process the same file twice.
proposed behavior: updates are still picked up asap but we might process the same version potentially twice.

(none of these are ideal - the question is whether we care about these cases more or less)

Reply:
Great summary. One point worth clarifying is the current behavior. Actually, the same version of the file may or may not be processed twice. After listing the file, the batch is processed one by one. The next file of the same batch could overwrite sincedb which gives an older timestamp, hence in the next iteration, the updated file is picked up again.

So, current and proposed behavior might process twice. I could imagine the updated s3 file which always matches cutoff time is unlikely to happen. Looking over the open issues, the common use-case is read and forget. Users seem not interest in the update part. I am happy to change it to process in the next iteration. Thanks to your input @kares  😃

Reply:
Thanks Kasie, appreciate the added logging - just in case a user does some unexpected updates we'll be able to better reason about what was going on from the log.

Would have been also fine to not care about the use-case as you proposed but in that case not have this as a patch release but rather a minor (just to be "safe").